File: Engine/src/main/java/org/etieskrill/engine/entity/service/impl/AnimationService.java
package org.etieskrill.engine.entity.service.impl;

import org.etieskrill.engine.entity.Entity;
import org.etieskrill.engine.entity.component.Drawable;
import org.etieskrill.engine.entity.service.Service;
import org.etieskrill.engine.graphics.animation.Animator;

import java.util.List;
import java.util.Set;

public class AnimationService implements Service {

    @Override
    public boolean canProcess(Entity entity) {
        return entity.hasComponents(Drawable.class, Animator.class);
    }

    @Override
    public void process(Entity targetEntity, List<Entity> entities, double delta) {
        Drawable drawable = targetEntity.getComponent(Drawable.class);
        Animator animator = targetEntity.getComponent(Animator.class);

        animator.update(delta);

        if (drawable.getShader() != null) {
            drawable.getShader().setUniformArray("boneMatrices", animator.getTransformMatrices().toArray());
        }
    }

    @Override
    public Set<Class<? extends Service>> runBefore() {
        return Set.of(RenderService.class);
    }

}

File: Engine/src/main/java/org/etieskrill/engine/entity/service/impl/AnimationService.kt
package org.etieskrill.engine.entity.service.impl;

import org.etieskrill.engine.entity.Entity
import org.etieskrill.engine.entity.component.Drawable
import org.etieskrill.engine.entity.service.Service
import org.etieskrill.engine.graphics.animation.Animator
import kotlin.reflect.KClass

class AnimationService : Service {

    override fun canProcess(entity: Entity) = entity.hasComponents<Drawable, Animator>()

    override fun process(targetEntity: Entity, entities: List<Entity>, delta: Double) {
        val drawable = targetEntity.getComponent<Drawable>()!!
        val animator = targetEntity.getComponent<Animator>()!!

        animator.update(delta)

        drawable.shader?.setUniformArray("boneMatrices", animator.transformMatricesArray)
    }

    override val runBefore: Set<KClass<out Service>>
        get() = setOf(RenderService::class)

}
